### DISCUSSION TITLE
Get list of files on pull request merge

### DISCUSSION THREAD

**Discussion by addietrr** [community/community#26604] [Actions]
<p>If I have the following workflow to trigger on a pull request merge:</p>
<pre><code>name: deployment

on:
  pull_request:
    types:
      - closed

jobs:
  job1:
    runs-on: ubuntu-latest

    steps:
    - name: Check out
      uses: actions/checkout@v1

    - name: get list of files in PR
      if: github.event.pull_request.merged
      run: |
        ???
</code></pre>
<p>What is the best way to get a list of files in the PR?</p>

**Comment by weide-zhou** [comment]
<p>You can use rest api to get the file list of pull request. Please refer to the <a href="https://developer.github.com/v3/pulls/#list-pull-requests-files" rel="nofollow noopener">link</a> for more details.</p>
<p>Code sample as below, use jq to filter the filename.</p>
<pre><code>- name: list files
        run: |
          URL="https://api.github.com/repos/${GITHUB_REPOSITORY}/pulls/${{ github.event.pull_request.number }}/files"
          FILES=$(curl -s -X GET -G $URL | jq -r '.[] | .filename')
          echo $FILES
</code></pre>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/0/02a190626d0decbc1738538573ebb7e6073c4270.png" data-download-href="/uploads/short-url/nh6sCb1DF9BHooojuXMa31Hn0Y.png?dl=1" title="pullrequestfiles.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/52772761/181098648-fba85455-2401-4d6f-8372-041cb675150b.png" alt="pullrequestfiles.png" data-base62-sha1="nh6sCb1DF9BHooojuXMa31Hn0Y" width="690" height="221" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/0/02a190626d0decbc1738538573ebb7e6073c4270_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">pullrequestfiles.png</span><span class="informations">706×227 6.68 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by luchiago** [comment]
<p>Hi <a class="mention" href="/weide-zhou">@weide-zhou</a>. Your snippet was very helpful, but I’m having some trouble and I don’t understand why, could you help me?<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/0/0668290230eaed7668a2ee1ff776a5881df16756.png" data-download-href="/uploads/short-url/UG1FMPgzLLUjyDu780RCubMDu6.png?dl=1" title="error.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/30028621/181098650-06425a33-bf50-4289-9a16-5c6c96aaf9aa.png" alt="error.png" data-base62-sha1="UG1FMPgzLLUjyDu780RCubMDu6" width="690" height="197" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/0/0668290230eaed7668a2ee1ff776a5881df16756_2_690x197.png, https://aws1.discourse-cdn.com/github/original/2X/0/0668290230eaed7668a2ee1ff776a5881df16756.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/0/0668290230eaed7668a2ee1ff776a5881df16756.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/0/0668290230eaed7668a2ee1ff776a5881df16756_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">error.png</span><span class="informations">977×279 37.4 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi <a class="mention" href="/luchiago">@luchiago</a> ,</p>
<p>The error is due to no json file changes in your pull request, which caused command ’FILES=$(curl … | | grep -E ‘.(json)$’)‘ to be failed.</p>
<p>You can check in this way:</p>
<pre><code>- id: changes
        run: |
          URL="https://api.github.com/repos/${{ github.repository }}/pulls/${{ github.event.pull_request.number }}/files"
          FILES=$(curl -s -X GET -G $URL | jq -r '.[] | .filename')
          if echo $FILES | grep -q ".json"; then
            echo "json file changed!!"
          else
            echo "no json file changed!!"
          fi
</code></pre>
<p>Thanks.</p>

**Comment by luchiago** [comment]
<p>Thanks <a class="mention" href="/weide-zhou">@weide-zhou</a>. It worked</p>

**Comment by PfanP** [comment]
<aside class="quote group-github-partners" data-username="weide-zhou" data-post="2" data-topic="17226">
<div class="title">
<div class="quote-controls"></div>
 weide-zhou:</div>
<blockquote>
<p><code>FILES=$(curl -s -X GET -G $URL | jq -r '.[] | .filename')</code></p>
</blockquote>
</aside>
<p>Hi <a class="mention" href="/weide-zhou">@weide-zhou</a> Thanks for your reply. It’s very helpful but I’m getting error like<br>
<code>Cannot index string with string "filename"</code></p>
<p>Thanks<br>
Khanh</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/self-coding-crab">@self-coding-crab</a>,</p>
<p>I notice you parsed ‘.json’ in your jq filter, it should be ‘.filename’ from the curl command. I confirmed the script works on my side.</p>
<p>Basically you can check the curl command without jq, check the response body, and then use jq to filter the file.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/4/467410f7c11e80d1ceb63146e2fe4f5a62890657.png" data-download-href="/uploads/short-url/a3g5OwStBC7xD0LYpCwBWKIgmer.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/52772761/181098653-4ef41917-4e13-4cb7-91a6-fdf6889686c8.png" alt="image" data-base62-sha1="a3g5OwStBC7xD0LYpCwBWKIgmer" width="690" height="294" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/4/467410f7c11e80d1ceb63146e2fe4f5a62890657_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">902×385 19.4 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>If i checked the incorrect workflow, please let me know.</p>
<p>Thanks</p>

**Comment by ksrisurapaneni** [comment]
<p>Hi <a class="mention" href="/weide-zhou">@weide-zhou</a> ,</p>
<p>I am seeing the same error, below is the command I am running. I am using a private repo, do we have to include some sort of authentication in the curl command so it takes the GITHUB_TOKEN?</p>
<p><code>jq: error (at &lt;stdin&gt;:4): Cannot index string with string "filename"</code></p>
<pre><code class="lang-auto">URL="https://api.github.com/repos/${{ github.repository }}/pulls/${{ github.event.pull_request.number }}/files"`

FILES=$(curl -s -X GET -G $URL | jq -r '.[] | .filename')
if echo $FILES | grep -q "Dockerfile"; then
  echo "Dockerfile file changed!!"
else
  echo "No Dockerfile file changed!!"
fi
</code></pre>
<p>Can you please help me figure out what I am doing wrong?</p>
<p>Thanks,<br>
Keerti</p>

**Comment by mattt** [comment]
<aside class="quote no-group" data-username="ksrisurapaneni" data-post="9" data-topic="17226">
<div class="title">
<div class="quote-controls"></div>
 ksrisurapaneni:</div>
<blockquote>
<p>I am using a private repo, do we have to include some sort of authentication in the curl command so it takes the GITHUB_TOKEN?</p>
</blockquote>
</aside>
<p>Yes, you’ll need to pass <code>GITHUB_TOKEN</code> to access non-public resources in your workflow. See<br>
“<a href="https://docs.github.com/en/actions/reference/authentication-in-a-workflow#using-the-github_token-in-a-workflow" rel="noopener nofollow ugc">Using the <code>GITHUB_TOKEN</code> in a workflow</a>” for more information.</p>
<pre><code class="lang-bash">URL="https://api.github.com/repos/${{ github.repository }}/pulls/${{ github.event.pull_request.number }}/files"`

FILES=$(curl -s -X GET -H "Authorization: Bearer ${{ secrets.GITHUB_TOKEN }}" $URL | jq -r '.[] | .filename')
if echo $FILES | grep -q "Dockerfile"; then
  echo "Dockerfile file changed!!"
else
  echo "No Dockerfile file changed!!"
fi
</code></pre>

**Comment by ksrisurapaneni** [comment]
<p>Hi <a class="mention" href="/mattt">@mattt</a> , thank you for responding. Your command works perfectly for pull events since the URL is taking info from a pull API. But it fails for push events or when we merge the PR.</p>
<p>So, I am using the below instead which works for both pull and push events:</p>
<aside class="onebox githubfolder">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/jitterbit/get-changed-files" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:400/400;"><img src="https://avatars.githubusercontent.com/u/19174065?s=400&amp;amp;v=4" class="thumbnail" width="400" height="400"></div>

<h3><a href="https://github.com/jitterbit/get-changed-files" target="_blank" rel="noopener nofollow ugc">jitterbit/get-changed-files</a></h3>


  <p><span class="label1">Get all of the files changed/modified in a pull request or push's commits.</span></p>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>Thank you,<br>
Keerti</p>
